elrond-codec-derive/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+    if fields.len() > 0 {
+        gen = quote! {
+            impl #impl_generics NestedDecode for #name #ty_generics #where_clause {
+                fn dep_decode<I: NestedDecodeInput>(input: &mut I) -> Result<Self, DecodeError> {
+                    Ok(#name {
+                        #(#fields: <#types>::dep_decode(input)?,)*
+                    })
+                }
+            
+                fn dep_decode_or_exit<I: NestedDecodeInput, ExitCtx: Clone>(
+                    input: &mut I,
+                    c: ExitCtx,
+                    exit: fn(ExitCtx, DecodeError) -> !,
+                ) -> Self {
+                    #name {
+                        #(#fields: <#types>::dep_decode_or_exit(input, c.clone(), exit),)*
+                    }
+                }
+            }
+        };
+    }
+    else {
+        gen = quote! {
+            impl #impl_generics NestedDecode for #name #ty_generics #where_clause {
+                fn dep_decode<I: NestedDecodeInput>(input: &mut I) -> Result<Self, DecodeError> {
+                    Ok(#name (
+                        #(<#types>::dep_decode(input)?),*
+                    ))
+                }
+            
+                fn dep_decode_or_exit<I: NestedDecodeInput, ExitCtx: Clone>(
+                    input: &mut I,
+                    c: ExitCtx,
+                    exit: fn(ExitCtx, DecodeError) -> !,
+                ) -> Self {
+                    #name (
+                        #(<#types>::dep_decode_or_exit(input, c.clone(), exit)),*
+                    )
+                }
+            }
+        };
+    }
+
+    gen.into()
+}
+
+// Top
+
+fn impl_top_encode_macro(ast: &syn::DeriveInput) -> TokenStream {
+    let name = &ast.ident;
+    let gen = match &ast.data {
+        syn::Data::Struct(_) => {
+            let (impl_generics, ty_generics, where_clause) = &ast.generics.split_for_impl();
+
+            quote! {
+                impl #impl_generics TopEncode for #name #ty_generics #where_clause {
+                    #[inline]
+                    fn top_encode<O: TopEncodeOutput>(&self, output: O) -> Result<(), EncodeError> {
+                        top_encode_from_nested(self, output)
+                    }
+                
+                    #[inline]
+                    fn top_encode_or_exit<O: TopEncodeOutput, ExitCtx: Clone>(
+                        &self,
+                        output: O,
+                        c: ExitCtx,
+                        exit: fn(ExitCtx, EncodeError) -> !,
+                    ) {
+                        top_encode_from_nested_or_exit(self, output, c, exit);
+                    }
+                }
+            }
+        },
+        syn::Data::Enum(_) => {
+            let idents = extract_field_names(&ast.data);
+            let value = 0..idents.len() as u8;
+            let name_repeated = std::iter::repeat(name);
+
+            quote! {
+                impl #name {
+                    pub fn to_u8(&self) -> u8 {
```

Comment:
I would avoid adding extra methods, to avoid clashes with user-defined methods.
The generated code can get duplicated, it's ok because we will never use top_encode in a contract.
I will play some more with traits, for now it can remain like that.

Reply:
Fixed.